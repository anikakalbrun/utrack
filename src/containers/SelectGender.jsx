import React, {Component} from 'react';
import {string, func} from 'prop-types';

import FilterWrapper from '../components/FilterWrapper';
import Button from '../components/Button';

const getBtnClass = (value, btnText) => value === btnText ? "btn-success" : "btn-light";
const isBtnDisabled = (btnClass) => btnClass === "btn-success";

class SelectGender extends Component {

    static propTypes = {
        value: string,
        setGenderFilter: func
    };

    static defaultProps = {
        value: 'All'
    };

    handleClickMaleBtn = () => {
        this.props.setGenderFilter('Male');
    };

    handleClickFemaleBtn = () => {
        this.props.setGenderFilter('Female');
    };

    handleClickAllBtn = () => {
        this.props.setGenderFilter('All');
    };

    render() {
        const {value} = this.props;
        return (
            <FilterWrapper label='Gender'>
                <div className='empty'/>
                <div className="btn-group" role="group" aria-label="gender">
                    <Button text='Male'
                            btnClass={getBtnClass(value, 'Male')}
                            disabled={isBtnDisabled(getBtnClass(value, 'Male'))}
                            handleClick={this.handleClickMaleBtn}/>
                    <Button text='Female'
                            btnClass={getBtnClass(value, 'Female')}
                            disabled={isBtnDisabled(getBtnClass(value, 'Female'))}
                            handleClick={this.handleClickFemaleBtn}/>
                    <Button text='All'
                            btnClass={getBtnClass(value, 'All')}
                            disabled={isBtnDisabled(getBtnClass(value, 'All'))}
                            handleClick={this.handleClickAllBtn}
                    />
                </div>
            </FilterWrapper>
        )
    }
}

export default SelectGender