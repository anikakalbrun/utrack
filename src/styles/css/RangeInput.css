.value {
    text-align: center;
    margin-top: 10px;
}