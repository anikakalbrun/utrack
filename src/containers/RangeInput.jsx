import React, {Component} from 'react';
import Slider from 'react-rangeslider';
import 'react-rangeslider/lib/index.css';
import '../styles/css/RangeInput.css';
import throttle from 'lodash.throttle';
import {number, func} from 'prop-types';

import FilterWrapper from '../components/FilterWrapper';

export const minValue = 0;
export const maxValue = 5000;
export const labels = {
    [minValue]: `$${minValue}`,
    [maxValue]: `$${maxValue}`
};
export const format$ = value => `$${value}`;

class RangeInput extends Component {

    static propTypes = {
        value: number,
        setSpendFilter: func.isRequired
    };

    static defaultProps = {
        value: 0
    };

    handleChange = (value) => {
        //only invoke this.props.setSpendFilter at most once per every 300 milliseconds
        const throttled = throttle(this.props.setSpendFilter, 300);
        throttled(value);
    };

    render() {
        const {value} = this.props;
        return (
            <FilterWrapper label={'Minimal spend'}>
                <div className='value'>{format$(value)}</div>
                <Slider
                    value={value}
                    onChange={this.handleChange}
                    min={minValue}
                    max={maxValue}
                    formatLabel={format$(value)}
                    format={format$}
                    labels={labels}
                />
            </FilterWrapper>
        )
    }
}

export default RangeInput