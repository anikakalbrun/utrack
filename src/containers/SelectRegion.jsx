import React, {Component} from 'react';
import Select from 'react-select';
import 'react-select/dist/react-select.css';
import {string, func} from 'prop-types';

import FilterWrapper from '../components/FilterWrapper';

export const options = [
    {value: 'United States', label: 'United States'},
    {value: 'Europe', label: 'Europe'},
    {value: 'APAC', label: 'APAC'},
    {value: 'Latin America', label: 'Latin America'},
];

class SelectRegion extends Component {

    static propTypes = {
        value: string,
        setRegionFilter: func.isRequired
    };

    static defaultProps = {
        value: ''
    };

    handleChange = (selectedOption) => {
        selectedOption && selectedOption.value ?
            this.props.setRegionFilter(selectedOption.value) :
            this.props.setRegionFilter('')
    };

    render() {
        const {value} = this.props;

        return (
            <FilterWrapper label={'Select region'}>
                <div className='empty'/>
                <Select
                    name="region"
                    lebel='region'
                    value={value}
                    onChange={this.handleChange}
                    options={options}
                    placeholder='Select region'
                />
            </FilterWrapper>
        );
    }
}

export default SelectRegion