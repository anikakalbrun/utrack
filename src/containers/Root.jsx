import React, {Component} from 'react';
import {connect} from "react-redux";
import {string, object, func, array, number} from 'prop-types';

import {
    getData,
    setSpendFilter,
    setRegionFilter,
    setGenderFilter
} from '../actions/index'
import {
    filteredUsersSelector,
    spendFilterSelector,
    regionFilterSelector,
    genderFilterSelector,
    chartDataSelector
} from '../reducers/index';
import InputRange from './RangeInput';
import SelectRegion from './SelectRegion';
import SelectGender from './SelectGender';
import DiscountCostLineChart from '../components/DiscountCostLineChart';
import UserTable from '../components/UserTable';
import Row from '../components/Row';

export class Root extends Component {

    static propTypes = {
        getData: func.isRequired,
        data: array.isRequired,
        spendFilter: number.isRequired,
        setSpendFilter: func.isRequired,
        regionFilter: string.isRequired,
        setRegionFilter: func.isRequired,
        genderFilter: string.isRequired,
        setGenderFilter: func.isRequired,
        chartData: object.isRequired
    };

    componentDidMount() {
        //generates the application data
        //getData takes as argument the number of records that should be generated
        this.props.getData(1000);
    }

    render() {
        const {
            data,
            spendFilter,
            setSpendFilter,
            regionFilter,
            setRegionFilter,
            genderFilter,
            setGenderFilter,
            chartData
        } = this.props;

        return (
            <div className='container'>
                <Row>
                    <InputRange value={spendFilter} setSpendFilter={setSpendFilter}/>
                    <SelectRegion value={regionFilter} setRegionFilter={setRegionFilter}/>
                    <SelectGender value={genderFilter} setGenderFilter={setGenderFilter}/>
                </Row>
                <Row>
                    <DiscountCostLineChart data={chartData}/>
                </Row>
                <Row>
                    <UserTable data={data}/>
                </Row>
            </div>
        )
    }
}

export function mapStateToProps(state) {
    return {
        data: filteredUsersSelector(state),
        chartData: chartDataSelector(state),
        spendFilter: spendFilterSelector(state),
        regionFilter: regionFilterSelector(state),
        genderFilter: genderFilterSelector(state)
    }
}

export default connect(mapStateToProps, {getData, setSpendFilter, setRegionFilter, setGenderFilter})(Root)