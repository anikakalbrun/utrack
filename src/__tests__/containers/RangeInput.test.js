import React from 'react';
import RangeInput from '../../containers/RangeInput';
import {minValue, maxValue, labels, format$} from '../../containers/RangeInput';

let wrapper;
const setSpendFilter = jest.fn();
const props = {
    setSpendFilter,
    value: 12
};

describe('RangeInput', () => {

    beforeEach(() => {
        wrapper = shallow(<RangeInput {...props}/>);
    });

    it('renders FilterWrapper', () => {
        expect(wrapper.find('FilterWrapper').length).toEqual(1);
    });

    it('passes the label property to FilterWrapper', () => {
        expect(wrapper.find('FilterWrapper').props().label).toEqual('Minimal spend');
    });

    it('renders selected value displayed in a div', () => {
        expect(wrapper.find('div.value').html()).toMatchSnapshot();
    });

    it('renders Slider', () => {
        expect(wrapper.find('Slider').length).toEqual(1);
    });

    it('passes proper properties to Slider', () => {
        expect(wrapper.find('Slider').props().value).toEqual(12);
        expect(wrapper.find('Slider').props().onChange).toBeTruthy();
        expect(wrapper.find('Slider').props().formatLabel).toEqual(format$(12));
        expect(wrapper.find('Slider').props().min).toEqual(minValue);
        expect(wrapper.find('Slider').props().max).toEqual(maxValue);
        expect(wrapper.find('Slider').props().labels).toEqual(labels);
    });

    it('changing sliders value triggers the setSpendFilter action creator', () => {
        wrapper.find('Slider').simulate('change');
        expect(setSpendFilter).toHaveBeenCalled();
    });
});
