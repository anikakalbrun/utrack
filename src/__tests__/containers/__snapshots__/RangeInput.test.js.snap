// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RangeInput renders selected value displayed in a div 1`] = `"<div class=\\"value\\">$12</div>"`;
